docs/error-recovery-ideas.md

```diff
@@ -0,0 +1,91 @@
+<!-- markdown-toc start - Don't edit this section. Run M-x markdown-toc-refresh-toc -->
+**Table of Contents**
+
+- [GetToken() doesn't indicate success/failure](#gettoken-doesnt-indicate-successfailure)
+- [Out-of-sync closers](#out-of-sync-closers)
+- [Dual levels of Error verbosity](#dual-levels-of-error-verbosity)
+- [Filtering Cascaded Errors](#filtering-cascaded-errors)
+- [Documenting how Error recover works](#documenting-how-error-recover-works)
+
+<!-- markdown-toc end -->
+
+Introduction
+------------
+
+This is a to-do list and a place for discussion of how to improve error handling, and may guide current development.
+
+GetToken() doesn't indicate success/failure
+-------------------------------------------
+Routines that call _GetToken()_, like _expectIdentifierToken()_ can't take corrective action. Here is an example.
+
+In
+
+```
+    function f() public pure {
+        address payable;
+		//             ^ identifier expected here
+    }
+```
+
+there is a semicolon where an identifier is expected. _expectIdentifierToken()_ will gobble this up leaving no semicolon for _Statement_ to synchronize to. for _expectIdentifierToken()_ to be smarter, it would need to get status back from _GetToken()_.
+
+Out-of-sync closers
+-------------------
+
+Steve C Johnson recovery is a coarse kind of recovery, and here it can cause cascaded effects as a result of skipping over one kind of terminator while looking for another.
+Right now, there is only one synchronizing token at a time for a rule,
```

Comment:
Why is that the case? It sounds to me like a rather good strategy to accumulate sync tokens in a kind of stack.

While we are at the statement level, we can synchronize to the next statement by searching for `;` or `}`, whichever comes first.

While we are inside a parenthesized expression (function call arguments, parameter list, ...), searching for `)` or `,` should do the trick. As soon as we encounter any of `{`, `;` or `}`, we should immediately throw an exception that is then caught at the statement level.

Reply:
Is this what you describe below?

Reply:
> Why is that the case? 

I was just trying to describe what the S C Johnson method is, because it is largely what has been implemented. I was not trying to suggest it is ideal. Rather, that it is a simple well-trodden place to start, and the kind of thing done here is not just some hacky code pulled out of the air.

> It sounds to me like a rather good strategy to accumulate sync tokens in a kind of stack.

Yes. I am in favor of that! I am suggesting to defer doing this after we have some sort of integration with an IDE or front-end tool.  The motivation for writing this up is a concern that people might get the misimpression that because there are things that aren't done right now in PR's that it was not in my mind or not my intention to complete later. 

The last pull request was too large, so I am attempting to reduce that by simplifying the error recovery and scope at the outset. 

However once this moves forward,  and we have some basic test cases and examples, I think it easy to see the specific benefit, and show how this improves on SC Johnson error handling in very specific ways.